tracing: Move custom-tag application after common tags

matthewgrossman opened the pull request:
resolves  https://github.com/envoyproxy/envoy/issues/20635. 

I can't imagine this matters, but when compiling I had to use `bazel build source/exe:envoy --define=wasm=disabled` to get it running on my M1 Max processor (thanks @keith!). I don't think this would make the code/testing less comprehensive, but figured I'd let reviewers know
<!--
!!!ATTENTION!!!

If you are fixing *any* crash or *any* potential security issue, *do not*
open a pull request in this repo. Please report the issue via emailing
envoy-security@googlegroups.com where the issue will be triaged appropriately.
Thank you in advance for helping to keep Envoy secure.

!!!ATTENTION!!!

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/main/PULL_REQUESTS.md)
-->

Commit Message: Ensure that configurable custom_tags are applied after the common tags. This guarantees that user configuration takes precedence over the hard-coded tags that envoy sets automatically in `setCommonTags`
Additional Description: n/a
Risk Level: Low
Testing: Unit testing
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: n/a
[Optional Runtime guard:] n/a
[Optional Fixes #Issue] https://github.com/envoyproxy/envoy/issues/20635
[Optional Fixes commit #PR or SHA] n/a
[Optional Deprecated:] n/a
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):] n/a

matthewgrossman commented:
sorry, PR is still in draft, not sure what my rebase + force-push did to tag a bunch of people 🙈

mengmichael1 commented on test/common/tracing/http_tracer_impl_test.cc:
```diff
@@ -865,6 +866,37 @@ TEST_F(HttpTracerImplTest, MetadataCustomTagReturnsDefaultValue) {
+  {
+    // ensure our setTag(..., "override_component") happens later, taking precedence
+    InSequence s;
+    EXPECT_CALL(span, setTag(Eq(custom_tag.tag()), "proxy"));
```
```suggestion
    EXPECT_CALL(span, setTag(Eq(custom_tag.tag()), Eq(Tracing::Tags::get().Proxy)));
```

matthewgrossman commented:
@wrowe @markdroth are you the right reviewers to be assigned to this PR, or did that happen accidentally due to my rebase flagging certain files under your domain?

if not, should we change the reviewers? cc @Shikugawa , who commented on the original issue https://github.com/envoyproxy/envoy/issues/20635

Shikugawa commented on source/common/tracing/http_tracer_impl.cc:
```diff
@@ -206,6 +206,11 @@ void HttpTracerUtility::finalizeDownstreamSpan(Span& span,
       addGrpcRequestTags(span, *request_headers);
     }
   }
+  span.setTag(Tracing::Tags::get().RequestSize, std::to_string(stream_info.bytesReceived()));
```
Could we move these `RequestSize` and `ResponseSize` inside `setCommonTags`?

matthewgrossman replied:
(fyi: the only reason the diff shows the addition of `RequestSize` and `ResponseSize` is because I moved them up above the `CustomTagContext` loop; they were already in this function)

happy to move it, but worried it might be intentionally left out of `setCommonTags`? `setCommonTags` is called in both:
- `finalizeDownstreamSpan()`
- `finalizeUpstreamSpan()`

the `RequestSize`/`ResponseSize` are only set as tags in `finalizeDownstreamSpan`, so adding it to common-tags would put it in `finalizeUpstreamSpan` as well. Does this make sense for that method? Would those values be set?

matthewgrossman replied:
here's where the `bytesSent()` bit was added: https://github.com/envoyproxy/envoy/pull/8289/files#diff-36ebe2d9da39e88ceeccb197a9ff0b8a1dc97a5f72c8597c40d27f61a9ead00fR176

I guess I'd be concerned about a comment like this: https://github.com/envoyproxy/envoy/pull/8289#discussion_r326426387, where certain fields are only set up for Downstream, but not Upstream spans, for example

matthewgrossman replied:
ah I think its fine, followed the breadcrumbs and I don't think we need to worry about that

Shikugawa replied:
@matthewgrossman The change to move `RequestSize` and `ResponseSize` might be out of range of this PR. For now It's OK not to reflect this change. Sorry for the noise ;)

matthewgrossman replied:
all good, was optimistic it wouldn't require too much but there's maybe slightly more there than first expected. Or just requires a simple change ¯\_(ツ)_/¯


removed the code and tests that were validating the new location

Shikugawa reviewed (commented):
Generally LGTM with nit comment. Thanks!

Shikugawa commented:
And could you check DCO signoff?

Shikugawa reviewed (dismissed):
LGTM in code. Take over @wbpcode

wbpcode commented:
Hi, thanks for your contributions. It's LGTM. Can you check your DCO? Looks like only the latest commit lose the sign. 😄

wbpcode commented:
/wait

matthewgrossman commented:
urgh sorry thought I got the DCO, fixing now

matthewgrossman commented:
Okay, fixed the DCO and some tests that were broken by moving request/response tags into `setCommonTags`. Will make sure tests are passing, but then think things are looking good

wbpcode commented:
Thanks for you updates. But it seems that the tests still failed. Can you check it again?

/wait

matthewgrossman commented:
The changes causing the failure were due to how I modified the request/response tags to be in `setCommonTags`, so now both upstream and downstream spans would see those emitted. Removed those changes to reduce noise in this PR and pushed things back up. Will monitor as tests run today

matthewgrossman commented:
looks like a timeout in the `envoy-presubmit` job, doubt it was anything related to this PR? will push a noop commit to trigger a fresh CI run

matthewgrossman commented:
/retest

wbpcode reviewed (approved):
LGTM. Thanks @matthewgrossman. 👍
